keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,480 @@
+since that mechanism still doesn't exist for built-in. This proposal
+tries to solve that gap.
+
+## Motivation
+
+Having a new `// +default` marker for built-in types has several
+important benefits:
+
+- It's more convenient for types authors.
+- The types and their behaviors are easier to understand since they are
+described together.
+- They are less error-prone, and prevent people from using anti-patterns
+like "cross-field dependency defaults" (default depends on another
+field), non-deterministic or complicated defaults. These patterns
+might be useful in limited circumstances, but it should be easier to do
+the right thing in most cases.
+- Removes the inconsistency with CRDs that already support this
+mechanism.
+- Solves bug where associative keys need to either be required or
+  defaulted.
+
+### Goals
+
+The goal is to add a new `// +default` tag to our current built-in Go
+IDL. That tag will be transformed into the OpenAPI `default` tag and
+then routed to defaulting functions so that defaulting can be done
+declaratively.
+
+### Non-Goals
+
+While this proposal is probably paving the way for additional OpenAPI
+based validation, validation is entirely out-of-scope for this project.
+
+This proposal also doesn't intend to remove the existing defaulting
+mechanism since some defaulting still requires non-static values that
+need to be computed and can't be defined declaratively. We could replace
+some defaulting that doesn't require to be done as code though.
+
+## Proposal
+
+Since the goal of this proposal is to increase consistency with the CRD
+defaulting, we propose to use the same mechanism (both in OpenAPI and
+similar to what kubebuilder does). This defaulting will happen BEFORE
+the existing defaulting functions are called (existing functions can be
+updated to panic so that we can see if defaults are properly applied).
+
+One very important nuance with CRDs is that CRDs are deserialized in
+unstructured objects that can distinguish between specified and
+unspecified fields. That is not the case for built-in types that are
+directly deserialized into Go-struct, where the unspecified fields are
+zero-valued. It's important to understand that only zero-value
+omit-empty fields will be defaulted using that mechanism.
+
+Computed defaults will not be supported as a declarative marker and will
+continue being supported through the existing defaulting functions.
+
+### Risks and Mitigations
+
+The main risk is that we could end-up with different defaults for
+existing types which would be extremely confusing for our users.
+
+## Design Details
+
+There are multiple challenges to running defaults declaratively and
+changing the current mechanisms:
+
+- We don't use OpenAPI to drive validation nor defaulting
+currently. There is no routing of the OpenAPI to most operations in
+kubernetes (the only thing right now is server-side apply).
+- OpenAPI fields don't strictly map to Go types fields
+(e.g. capitalization, embedding)
```

Comment:
what does this mean? Coming back to my question about reflection. I don't feel week with introducing a reflection based mechanism. We just spent tons of time to remove reflection based conversion from the stack. We probably want to avoid to add a new step in our pipeline.

Reply:
nvmd. You come to that below.

In general, it is much easier to follow if you clearly propose first what you want to do and then afterwards discuss it.